docs/HowToFAQ/FAQLoopModifies.md

```diff
@@ -0,0 +1,90 @@
+This saves users from having to write loop invariants about all other variables.
+Only invariants about variables that are modified are needed.
+
+What about the heap? For this purpose, the heap is considered to be one variable. 
+So, the heap is either an assignment target of the loop or it isn’t. 
+This means that if your loop body assigns to anything in the heap, the heap becomes an assignment target. 
+Also, if the loop body allocates another object, that too is a change of the heap, 
+so the heap becomes an assignment target. 
+Furthermore, Dafny’s rule about a method is that a method is allowed to change the 
+state of any object in the method’s modifies clause, 
+and it is also allowed to allocate new objects and to modify their state. 
+Therefore, if your loop body calls a method, then the heap may well become an assignment target.
+
+So, then, does this mean your loop invariant needs to speak about everything in the heap whenever the heap is an assignment target of the loop?
+
+Again, no. Loops in Dafny have `modifies` clauses, just like methods do (and just like for methods, 
+newly allocated objects are implicitly included in those `modifies` clauses). 
+It would be a pain to even have to write `modifies` clauses for every loop, 
+so if a loop does not have an explicit `modifies` clause, Dafny implicitly attaches 
+the `modifies` clause of the enclosing method (or of the enclosing loop, in the case of nested loops). 
+This default behavior turns out to be right in almost all cases in practice, 
+which is why you normally don’t have to think about this issue.
+
+But when it is not, you need to be explicit about the `modifies` clause and the loop invariants. 
+In particular
+- write a `modifies` clause for the loop that includes all objects whose fields might be 
+assigned in the loop body (or listed in the modifies clauses of anything the loop body calls)
+- then write a loop invariant that includes assertions about any variable that is listed in the modifies clause
+or is an assignment target in the loop body. 
+Very typically, the invariant will give a value for a havoced variable showing its relationship to the loop index.
+
+For example, a loop that sets the elements of an array to some initial value might look like this:
+```dafny
+method init(a: array<int>) 
+  modifies a
+  ensures forall j | 0 <= j < a.Length :: a[j] == j;
+{
+  var i: int := 0;
+
+  while (i < a.Length) 
+    modifies a
+    invariant 0 <= i <= a.Length && forall j | 0 <= j < i :: a[j] == j;
+  {
+    a[i] := i;
+    i := i+1;
+  }
+}
+```
+
+Note the following points:
+- The method specifies that it might modify the elements of the array `a`.
+- The loop says that it modifies the same array. 
+In fact that modifies clause could be omitted
+because it would be inherited from the enclosing context
+- The loop also modifies `i`, but as `i` is a local variable of non-reference type, it need not be listed in the modifies clause.
```

Comment:
Omit "of non-reference type". What is said holds for local variables of any type. (I think the comment is trying to get to that the _state_ referenced by a reference-type local variable may or may not be modified. But I think that's confusing to bring up here.)

Reply:
Edited as you suggest. Though I think I would be confused. Are you saying that a local variable of reference type nevoid not be in a modifies clause even if its state is changed?

Reply:
All `modifies` clauses, for both methods and loops, always only talk about which parts of the heap are modified. If you want to modify a local variable in a loop, you just go ahead and do so--no specification needed, because Dafny will peer into the loop body and figure out which local variables may be modified.